### DISCUSSION TITLE
Help !

### DISCUSSION THREAD

**Discussion by AureliaDolo** [adrien-zinger/wasmer-as#2] [Q&A] [answered]
There is still these functions to implement
* [ ] pin
* [ ] unpin
* [ ] collect
* [ ] write when the size is different

I have a few question : 
- how do you know it's this functions ? (where is the doc?)
- they will be passed to Env::new() but where should they be implemented ?
- you had an example of an implementation of this functions, where is it ?

**Comment by adrien-zinger** [comment, marked as answer]
Hey !

Here is the documentation of what is exported by AssemblyScript -> [https://www.assemblyscript.org/garbage-collection.html#runtime-interface](https://www.assemblyscript.org/garbage-collection.html#runtime-interface)

[The new is implemented here](https://github.com/adrien-zinger/wasmer-as/blob/8ab64392ac8ad1a70243ef494e9cd5a187df8602/src/env.rs#L9-L16) is only used when you need the env outside the ABI (in the test for exemple), you'll have to add a parameter in the new :+1: check in the [unit test](https://github.com/adrien-zinger/wasmer-as/blob/8ab64392ac8ad1a70243ef494e9cd5a187df8602/tests/strings.rs#L49-L55).

But when it's in the ABI, wasmer call automatically [this](https://github.com/adrien-zinger/wasmer-as/blob/8ab64392ac8ad1a70243ef494e9cd5a187df8602/src/env.rs#L18-L31). Here you can do the same than in the unit test to get the function.

> We can may be use `#[cfg(test)]` for the implementation of the new? :thinking:

**Reply by adrien-zinger** [reply]
It's a little bit hard core to read this :rofl: but you can see in the .wat this for the new... if you're interested of how the runtime work you can follow the calls starting with the function `$~lib/rt/itcms/__new` :robot: 

https://github.com/adrien-zinger/wasmer-as/blob/8ab64392ac8ad1a70243ef494e9cd5a187df8602/tests/runtime_exported.wat#L47

**Reply by AureliaDolo** [reply]
Thanks ! :heart:

**Reply by adrien-zinger** [reply]
You can also see the javascript implementation in the [AssemblyScript project](https://github.com/AssemblyScript/assemblyscript).
For exemple, the "__newString" already implemented is inspired of [this](https://github.com/AssemblyScript/assemblyscript/blob/8fe5248a22f59b23b381243b1ea3ed92f5b4c222/lib/loader/index.js#L144-L151):
```javascript
  function __newString(str) {
    if (str == null) return 0;
    const length = str.length;
    const ptr = __new(length << 1, STRING_ID);
    const U16 = new Uint16Array(memory.buffer);
    for (var i = 0, p = ptr >>> 1; i < length; ++i) U16[p + i] = str.charCodeAt(i);
    return ptr;
  }
```

**Reply by AureliaDolo** [reply]
Thanks a lot !
